src/main/scala/com/advancedtelematic/director/db/Repository.scala

```diff
@@ -156,17 +157,39 @@ protected class AssignmentsRepository()(implicit val db: Database, val ec: Execu
     Schema.assignments.filter(_.deviceId === deviceId).result
   }
 
+  def findValidBy(deviceId: DeviceId): Future[Seq[Assignment]] = db.run {
+    Schema.assignments.filter(_.deviceId === deviceId).filterNot(_.status === AssignmentStatus.Canceled).result
+  }
+
   def existsFor(ecus: Set[EcuIdentifier]): Future[Boolean] = db.run {
     Schema.assignments.filter(_.ecuId.inSet(ecus)).exists.result
   }
 
   def findLastCreated(deviceId: DeviceId): Future[Option[Instant]] = db.run {
-    Schema.assignments.filter(_.deviceId === deviceId).sortBy(_.createdAt.reverse).map(_.createdAt).result.headOption
+    Schema.assignments.filter(_.deviceId === deviceId).filterNot(_.status === AssignmentStatus.Canceled)
+      .sortBy(_.createdAt.reverse).map(_.createdAt).result.headOption
   }
 
   def setAllInFlight(deviceId: DeviceId): Future[Unit] = db.run {
-    Schema.assignments.filter(_.deviceId === deviceId).map(_.inFlight).update(true).map(_ => ())
+    Schema.assignments
+      .filter(_.deviceId === deviceId)
+      .filter(_.status === AssignmentStatus.Pending)
+      .map(_.status).update(InFlight)
+      .map(_ => ())
+  }
+
+  def setCanceled(ns: Namespace, deviceIds: Seq[DeviceId]): Future[Seq[Assignment]] = {
```

Comment:
Canceled assignments must be moved to processed assignments table, otherwise the device wont be able to get a new assignment after one assignment is canceled? Pk for assignments is just device id? We should probably have a test for that:

1. add assignment for device
2. cancel assignment
3. device doesn't see new assignment
4. add assignment
5. device sees assignment

Reply:
Ah ok, I thought "ProcessedAssignments" means "processed by the device". I did filter out the canceled assignments at places, but it probably makes more sense to move them. The ProcessedAssignments need a "cancelled" flag then, right?

Btw, why does `AssignmentsRepository.findBy()` not take the namespace into account? Are there situations where we talk about a device outside of a namespace?

Reply:
The idea with `ProcessedAssignments` was for it to represent "processed" on the server side, so I think would make sense to have a flag/status then.

I was trying to avoid passing the namespace everywhere, because the deviceId should be universally unique, and the namespace is more of an authorization concern that should be checked at the resource level IMO. That would simplify code downstream, but would introduce some risk if we don't verify the authorization upstream.